src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -125,9 +120,44 @@ create_distributed_table(PG_FUNCTION_ARGS)
 	text *distributionColumnText = PG_GETARG_TEXT_P(1);
 	Oid distributionMethodOid = PG_GETARG_OID(2);
 
+	text *colocateWithTableNameText = PG_GETARG_TEXT_P(3);
+	char *colocateWithTableName = text_to_cstring(colocateWithTableNameText);
+
 	char *distributionColumnName = text_to_cstring(distributionColumnText);
 	char distributionMethod = LookupDistributionMethod(distributionMethodOid);
 
+	/* check if we try to colocate with hash distributed tables */
+	if (strncmp(colocateWithTableName, "default", NAMEDATALEN) != 0 &&
+		strncmp(colocateWithTableName, "none", NAMEDATALEN) != 0)
+	{
+		bool colocateWithSupported = true;
+		Oid colocateWithTableOid = InvalidOid;
+		char colocateWithTablePartitionType = 0;
+
+		/* check the distribution type of the new table  */
+		if (distributionMethod != DISTRIBUTE_BY_HASH)
```

Comment:
Minor and it is up to you: Should we postpone this check after finding colocateWithTablePartitionType? So that you can eliminate colocateWithSupported variable completely. It will have small performance penalty only if distributionMethod != DISTRIBUTE_BY_HASH and table creation is not that frequent operation.